etl/steps/data/garden/tuberculosis/2023-11-27/burden_disaggregated.py

```diff
@@ -0,0 +1,183 @@
+    # Load income groups dataset.
+    ds_income_groups = paths.load_dependency("income_groups")
+    #
+    # Process data.
+    #
+    tb = geo.harmonize_countries(df=tb, countries_file=paths.country_mapping_path)
+    tb = tb.drop(columns=["measure", "unit"])
+    tb = add_population_column(tb)
+    tb = combining_sexes_for_all_age_groups(tb)
+    tb = add_region_sum_aggregates(tb, ds_regions=ds_regions, ds_income_groups=ds_income_groups)
+    tb, tb_rate = calculate_incidence_rates(tb)
+    tb = tb.set_index(["country", "year", "age_group", "sex", "risk_factor"], verify_integrity=True)
+    tb_rate = tb_rate.set_index(["country", "year", "age_group", "sex", "risk_factor"], verify_integrity=True)
+    #
+    # Save outputs.
+    #
+    # Create a new garden dataset with the same metadata as the meadow dataset.
+    ds_garden = create_dataset(
+        dest_dir, tables=[tb, tb_rate], check_variables_metadata=True, default_metadata=ds_meadow.metadata
+    )
+
+    # Save changes in the new garden dataset.
+    ds_garden.save()
+
+
+def combining_sexes_for_all_age_groups(tb: Table) -> Table:
+    """
+    Not all of the age-groups provided by the WHO have a value for both sexes, so we need to combine values for males and females to calculate these.
+    """
+
+    tb["age_group"] = tb["age_group"].astype("str")
+    age_groups_with_both_sexes = tb[tb["sex"] == "a"]["age_group"].drop_duplicates().to_list()
+    msk = tb["age_group"].isin(age_groups_with_both_sexes)
+    tb_age = tb[~msk]
+    tb_gr = (
+        tb_age.groupby(["country", "year", "age_group", "risk_factor"], dropna=False)[
+            ["best", "lo", "hi", "population"]
+        ]
+        .sum()
+        .reset_index()
+    )
+    tb_gr["sex"] = "a"
+    # Set population to nan for rows where the risk factor is not "all"
+    tb_gr.loc[tb_gr["risk_factor"] != "all", "population"] = np.nan
+    tb = pr.concat([tb, tb_gr], axis=0, ignore_index=True, short_name=paths.short_name)
+
+    return tb
+
+
+def add_region_sum_aggregates(tb: Table, ds_regions: Dataset, ds_income_groups: Dataset) -> Table:
+    """
+    Calculate region aggregates for all for each combination of age-group, sex and risk factor in the dataset.
+    """
+    # Create the groups we want to aggregate over.
+    tb_gr = tb.groupby(["year", "age_group", "sex", "risk_factor"])
+    tb_gr_out = Table()
+    for gr_name, gr in tb_gr:
+        for region in REGIONS_TO_ADD:
+            # List of countries in region.
+            countries_in_region = geo.list_members_of_region(
+                region=region,
+                ds_regions=ds_regions,
+                ds_income_groups=ds_income_groups,
+            )
+            gr_cal = gr[["country", "year", "best", "lo", "hi", "population"]]
+            # Add region aggregates.
+            gr_reg = geo.add_region_aggregates(
+                df=gr_cal,
+                region=region,
+                countries_in_region=countries_in_region,
```

Comment:
Unfortunately, you must pass here the argument `countries_that_must_have_data=[]`, otherwise it will impose that certain countries must be informed. This is one of the reasons we should create a new function to create region aggregates.

Reply:
Even worse than that, it will silently load `key_indicators` to achieve that, which is not listed as a dependency. So, to avoid this issue, simply add the `countries_that_must_have_data=[]`.

Reply:
Ah I didn't realise this before - thanks!